lib/transition/unpack.c

```diff
@@ -164,6 +164,130 @@ unpack_synapse(crm_graph_t * new_graph, xmlNode * xml_synapse)
     return new_synapse;
 }
 
+static xmlNode *
+get_start_action_xml(synapse_t *synapse)
+{
+    GListPtr lpc;
+
+    for (lpc = synapse->actions; lpc != NULL; lpc = lpc->next) {
+        crm_action_t *action = (crm_action_t *) lpc->data;
+        const char *operation = crm_element_value(action->xml, XML_LRM_ATTR_TASK);
+
+        if (action->type == action_type_rsc && safe_str_eq(operation, RSC_START)) {
+            return action->xml;
+        }
+    }
+
+    return NULL;
+}
+
+static int
+get_max_action_id(crm_graph_t *graph)
+{
+    GListPtr sIter;
+    int id = 0;
+
+    for (sIter = graph->synapses; sIter != NULL; sIter = sIter->next) {
+        synapse_t *synapse = (synapse_t *) sIter->data;
+        GListPtr aIter;
+
+        for (aIter = synapse->actions; aIter != NULL; aIter = aIter->next) {
+            crm_action_t *action = (crm_action_t *) aIter->data;
+
+            if (action->id > id) {
+                id = action->id;
+            }
+        }
+    }
+
+    return id;
+}
+
+static void destroy_action(crm_action_t * action);
+
+static gboolean
+inject_metadata(crm_graph_t *graph, synapse_t *parent_synapse)
+{
+    crm_action_t *new_input = NULL;
+    synapse_t *new_synapse = NULL;
+    xmlNode *synapse = NULL;
+    xmlNode *action_set = NULL;
+    xmlNode *op = NULL;
+    xmlNode *rsc = NULL;
+    xmlNode *attrs = NULL;
+    char *op_key = NULL;
+    xmlNode *start = NULL;
+    xmlNode *start_rsc = NULL;
+    xmlNode *start_attrs = NULL;
+
+    if (!graph || !parent_synapse) {
+        return FALSE;
+    }
+
+    start = get_start_action_xml(parent_synapse);
+    if (!start) {
+        return FALSE;
+    }
```

Comment:
> The versioned attributes may be needed for actions done before a start (such as the initial probe, or a stop resulting from a probe that detects a resource active where it shouldn't be).

I've only found the recovery scheduling for unexpectedly running remote resources in Pengine. This is done on the probe result unpacking. Because there is a short interval between a stop and a probe, could we use the metadata gained before the probe without performing a separate call?

> We somehow need to inject metadata before the first action of any kind on a resource, if and only if we don't already have metadata cached.

What do you mean by that? Besides the cases listed above, all of the actions are performed after the start. So, we have metadata in cache for any of them.